Add 2.5 blog (#11823)
prettier/prettier: website/blog/2021-11-21-2.5.0.md

// Prettier 2.4
SyntaxError: ';' expected. (1:36)
> 1 | import obj from "./something.json" assert { type: "json" };
    |                                    ^
  2 |

// Prettier 2.5
import obj from "./something.json" assert { type: "json" };

```

##### Handle `.mts` and `.cts`

Handle new file extensions `.mts` and `.cts`.

### MDX

#### Add support for MDX 2 comments ([#11563](https://github.com/prettier/prettier/pull/11563) by [@wooorm](https://github.com/wooorm))

This adds basic support for MDX 2 comments (JavaScript-style comments) in addition to existing MDX 1 comments (HTML-style comments).

Note that in MDX 2, more complex comments and whole expressions can be used inside braces.
This currently only supports a single comment on one line so that `{/* prettier-ignore */}` can be used, but doesn’t support most of MDX 2.

<!-- prettier-ignore -->
```mdx
Input:
{/*A comment*/}

Prettier 2.4:
{/_A comment_/}

Prettier 2.5:
{/*A comment*/}
```

## Other Changes

### JavaScript

#### Fix parentheses around sequence expression as body of arrow chain ([#11593](https://github.com/prettier/prettier/pull/11593) by [@bakkot](https://github.com/bakkot))

> The required parentheses around sequence expressions as the body of arrow functions were dropped for chained arrows, causing incorrect code to be generated. Now they are included.

Reviewer: ```suggestion
The required parentheses around sequence expressions as the body of arrow functions are now preserved for chained arrows. Previously, Prettier removed them, which resulted in invalid syntax.
```

<!-- prettier-ignore -->
```jsx
// Input
const f = () => () => (0, 1);

// Prettier 2.4
const f = () => () => 0, 1;

// Prettier 2.5
const f = () => () => (0, 1);
```

#### Ignore errors for sloppy mode syntax ([#11750](https://github.com/prettier/prettier/pull/11750) by [@fisker](https://github.com/fisker), [#11778](https://github.com/prettier/prettier/pull/11778) by [@sosukesuzuki](https://github.com/sosukesuzuki))

> 

Reviewer: ```suggestion

JavaScript’s [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) adds several useful errors to prevent mistakes. Some of these errors are syntax errors that occur at parse time. Since Prettier’s goal is to format all syntactically valid JavaScript code regardless of whether it will actually run, we’ve opted to leave this error checking to linters, compilers, and the runtime.

```

<!-- prettier-ignore -->
```jsx
// Input
function foo() { var bar = 1; delete bar; }

// Prettier 2.4
SyntaxError: Deleting local variable in strict mode. (1:31)
> 1 | function foo() { var bar = 1; delete bar; }
    |                               ^

// Prettier 2.5
function foo() {
  var bar = 1;
  delete bar;
}
```

#### Respect spacing for between expressions and parentheses in embedded CSS ([#11800](https://github.com/prettier/prettier/pull/11800) by [@sosukesuzuki](https://github.com/sosukesuzuki))

> <!-- Optional description if it makes sense. -->
> 
> <!-- prettier-ignore -->
> ```jsx
> // Input
> const paragraph2 = css`
>   transform: ${expr}(30px);
> `;
> 
> // Prettier 2.4
> const paragraph2 = css`
>   transform: ${expr} (30px);
> `;
> 
> // Prettier 2.5
> const paragraph2 = css`
>   transform: ${expr}(30px);
> `;
> 
> ```

Reviewer: ```suggestion
<!-- prettier-ignore -->
```jsx
// Input
const paragraph2 = css`
  transform1: ${expr}(30px);
  transform2: ${expr} (30px);
`;

// Prettier 2.4
const paragraph2 = css`
  transform1: ${expr} (30px);
  transform2: ${expr} (30px);
`;

// Prettier 2.5
const paragraph2 = css`
  transform1: ${expr}(30px);
  transform2: ${expr} (30px);
`;

```
```

#### Support ES2022 class-private-fields-in syntax in `espree` parser ([#11835](https://github.com/prettier/prettier/pull/11835) by [@fisker](https://github.com/fisker))

<!-- prettier-ignore -->
```jsx
// Input
class Foo {
  #brand;
  static isC(obj) {
    return #brand in Foo;
  }
}

// Prettier 2.4
SyntaxError: Unexpected token #brand (4:12)
  2 |   #brand;
  3 |   static isC(obj) {
> 4 |     return #brand in Foo;
    |            ^
  5 |   }
  6 | }

// Prettier 2.5
class Foo {
  #brand;
  static isC(obj) {
    return #brand in Foo;
  }
}

> ```

Reviewer: Similarly to the TypeScript thing I mentioned earlier:

```js
// Example
class Foo {
  #brand;
  static isC(obj) {
    return #brand in Foo;
  }
}
```

### TypeScript

#### Remove unnecessary parentheses for decorators ([#11717](https://github.com/prettier/prettier/pull/11717), [#11849](https://github.com/prettier/prettier/pull/11849) by [@sosukesuzuki](https://github.com/sosukesuzuki))

<!-- prettier-ignore -->
```tsx
// Input
class Test {
  @foo`bar`
  test1: string = "test"

  @test().x("global").y()
  test2: string = "test";
}

// Prettier 2.4
class Test {
  @(foo`bar`)
  test: string = "test"

  @(test().x("global").y())
  test2: string = "test";
}

// Prettier 2.5
class Test {
  @foo`bar`
  test: string = "test"

  @test().x("global").y()
  test2: string = "test";
}
```

### SCSS

> #### Improve `with (...)` formatting ([#11637](https://github.com/prettier/prettier/pull/11637) by [@sosukesuzuki](https://github.com/sosukesuzuki))

Reviewer: For consistency with the next entry:

```suggestion
#### Improve `@use with` formatting ([#11637](https://github.com/prettier/prettier/pull/11637) by [@sosukesuzuki](https://github.com/sosukesuzuki))
```

></div>

<!-- Prettier 2.4 -->
<div
  class="
    SomeComponent__heading-row
    d-flex
    flex-column flex-lg-row
    justify-content-start justify-content-lg-between
    align-items-start align-items-lg-center
  "
></div>

<!-- Prettier 2.5 -->
<div
  class="SomeComponent__heading-row d-flex flex-column flex-lg-row justify-content-start justify-content-lg-between align-items-start align-items-lg-center"
></div>

```

### Ember / Handlebars

#### Uses the opposite quote type for quotes inside mustache statements in attributes ([#11524](https://github.com/prettier/prettier/pull/11524) by [@bmaehr](https://github.com/bmaehr))

<!-- prettier-ignore -->
```hbs
{{!-- Input --}}
<div title="{{t 'login.username.description'}}" />

{{!-- Prettier 2.5 --}}
<div title="{{t 'login.username.description'}}" />

{{!-- Prettier 2.4 --}}
<div title="{{t "login.username.description"}}" />

```

### Markdown

#### Keep trailing commas for type parameters in embedded TSX ([#11685](https://github.com/prettier/prettier/pull/11685) by [@sosukesuzuki](https://github.com/sosukesuzuki))

> <!-- Optional description if it makes sense. -->

Reviewer: ```suggestion
The trailing comma is necessary to prevent TypeScript from treating the `<T>` as the beginning of a JSX expression. 
```
